File: python/gilded_rose.py
# -*- coding: utf-8 -*-

class GildedRose(object):

    def __init__(self, items):
        self.items = items
  
    def update_quality(self):
        for item in self.items:
            if item.name == "Sulfuras, Hand of Ragnaros":
                continue

            if item.name == "Aged Brie":
                self.update_aged_brie(item)
            elif item.name == "Backstage passes to a TAFKAL80ETC concert":
                self.update_backstage_passes(item)
            else:
                self.update_normal_item(item)

    def update_aged_brie(self, item):
        if item.quality < 50:
            item.quality += 1
        item.sell_in -= 1
        if item.sell_in < 0 and item.quality < 50:
            item.quality += 1

    def update_backstage_passes(self, item):
        if item.quality < 50:
            item.quality += 1
            if item.sell_in < 11:
                item.quality += 1
            if item.sell_in < 6:
                item.quality += 1
        item.sell_in -= 1
        if item.sell_in < 0:
            item.quality = 0

    def update_normal_item(self, item):
        if item.quality > 0:
            item.quality -= 1
        item.sell_in -= 1
        if item.sell_in < 0 and item.quality > 0:
            item.quality -= 1

class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)

File: php/src/GildedRose.php
<?php

declare(strict_types=1);

namespace GildedRose;

interface UpdateQualityStrategy
{
    public function updateQuality(Item $item): void;
}

class DefaultUpdateQualityStrategy implements UpdateQualityStrategy
{
    public function updateQuality(Item $item): void
    {
        if ($item->quality > 0) {
            $item->quality = $item->quality - 1;
        }
    }
}

class AgedBrieUpdateQualityStrategy implements UpdateQualityStrategy
{
    public function updateQuality(Item $item): void
    {
        if ($item->quality < 50) {
            $item->quality = $item->quality + 1;
        }
    }
}

class BackstagePassesUpdateQualityStrategy implements UpdateQualityStrategy
{
    public function updateQuality(Item $item): void
    {
        if ($item->quality < 50) {
            $item->quality = $item->quality + 1;
            if ($item->sellIn < 11 && $item->quality < 50) {
                $item->quality = $item->quality + 1;
            }
        }
    }
}

final class GildedRose
{
    private array $updateQualityStrategies;

    /**
     * @param Item[] $items
     */
    public function __construct(
        private array $items
    ) {
        $this->updateQualityStrategies = [
            'Aged Brie' => new AgedBrieUpdateQualityStrategy(),
            'Backstage passes to a TAFKAL80ETC concert' => new BackstagePassesUpdateQualityStrategy(),
            'default' => new DefaultUpdateQualityStrategy(),
        ];
    }

    public function updateQuality(): void
    {
        foreach ($this->items as $item) {
            $strategy = $this->updateQualityStrategies[$item->name] ?? $this->updateQualityStrategies['default'];
            $strategy->updateQuality($item);
        }
    }
}